datarobot_batch_scoring/writer.py

```diff
@@ -410,9 +405,45 @@ def format_result_data(self, result, batch, pred_name):
                            key=operator.itemgetter('row_id'))]
             out_fields = ['row_id', pred_name if pred_name else '']
         else:
-            ValueError('task "{}" not supported'
-                       ''.format(result['task']))
-        return (out_fields, pred)
+            raise ValueError('task "{}" not supported'
+                             ''.format(result['task']))
+
+        if keep_cols:
+            # stack columns
+
+            feature_indices = {col: i for i, col in
+                               enumerate(batch.fieldnames)}
+            indices = [feature_indices[col] for col in keep_cols]
+
+            written_fields = []
+
+            if not skip_row_id:
+                written_fields.append('row_id')
+
+            written_fields += keep_cols + out_fields[1:]
+
+            # first column is row_id
+            comb = []
+            for row, predicted in zip(batch.data, pred):
+                if fast_mode:
+                    # row is a full line, we need to cut it into fields
+                    # FIXME this will fail on quoted fields!
+                    row = row.rstrip().split(input_delimiter)
+                keeps = [row[i] for i in indices]
+                output_row = []
+                if not skip_row_id:
+                    output_row.append(predicted[0])
+                output_row += keeps + predicted[1:]
+                comb.append(output_row)
+        else:
+            if not skip_row_id:
+                comb = pred
+                written_fields = out_fields
+            else:
+                written_fields = out_fields[1:]
+                comb = [row[1:] for row in pred]
+
+        return (written_fields, comb)
 
     def sigterm_handler(self, _signo, _stack_frame):
```

Comment:
maybe it would make more sense in the main process so it can gracefully shutdown the other processes.